src/Lib.P2P/PeerManager.cs

```diff
@@ -0,0 +1,211 @@
+
+        /// <inheritdoc />
+        public Task StartAsync()
+        {
+            SwarmService.ConnectionEstablished += Swarm_ConnectionEstablished;
+            SwarmService.PeerNotReachable += Swarm_PeerNotReachable;
+
+            _cancel = new CancellationTokenSource();
+            var _ = PhoenixAsync(_cancel.Token);
+
+            _log.Debug("started");
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        public Task StopAsync()
+        {
+            SwarmService.ConnectionEstablished -= Swarm_ConnectionEstablished;
+            SwarmService.PeerNotReachable -= Swarm_PeerNotReachable;
+            DeadPeers.Clear();
+
+            _cancel.Cancel();
+            _cancel.Dispose();
+
+            _log.Debug("stopped");
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///   Indicates that the peer can not be connected to.
+        /// </summary>
+        /// <param name="peer"></param>
+        public void SetNotReachable(Peer peer)
+        {
+            var dead = DeadPeers.AddOrUpdate(peer,
+                new DeadPeer
+                {
+                    Peer = peer,
+                    Backoff = InitialBackoff,
+                    NextAttempt = DateTime.Now + InitialBackoff
+                },
+                (key, existing) =>
+                {
+                    existing.Backoff += existing.Backoff;
+                    existing.NextAttempt = existing.Backoff <= MaxBackoff
+                        ? DateTime.Now + existing.Backoff
+                        : DateTime.MaxValue;
+                    return existing;
+                });
+
+            SwarmService.BlackList.Add($"/p2p/{peer.Id}");
+            if (dead.NextAttempt != DateTime.MaxValue)
+            {
+                _log.DebugFormat("Dead '{0}' for {1} minutes.", dead.Peer, dead.Backoff.TotalMinutes);
+            }
+            else
+            {
+                SwarmService.DeregisterPeer(dead.Peer);
+                _log.DebugFormat("Permanently dead '{0}'.", dead.Peer);
+            }
+        }
+
+        /// <summary>
+        ///   Indicates that the peer can be connected to.
+        /// </summary>
+        /// <param name="peer"></param>
+        public void SetReachable(Peer peer)
+        {
+            _log.DebugFormat("Alive '{0}'.", peer);
+
+            DeadPeers.TryRemove(peer, out _);
+            SwarmService.BlackList.Remove($"/p2p/{peer.Id}");
+        }
+
+        /// <summary>
+        ///   Is invoked by the <see cref="SwarmService"/> when a peer can not be connected to.
+        /// </summary>
+        private void Swarm_PeerNotReachable(object sender, Peer peer) { SetNotReachable(peer); }
+
+        /// <summary>
+        ///   Is invoked by the <see cref="SwarmService"/> when a peer is connected to.
+        /// </summary>
+        private void Swarm_ConnectionEstablished(object sender, PeerConnection connection)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this unused method parameter 'sender'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)